rotary.py

```diff
@@ -7,107 +7,137 @@
     RANGE_WRAP = const(2)
     RANGE_BOUNDED = const(3)
- 
+
     def __init__(self, min_val, max_val, reverse, range_mode):
         self._min_val = min_val
         self._max_val = max_val
         self._reverse = -1 if reverse else 1
         self._range_mode = range_mode
         self._value = min_val
         self._state = _R_START
-        
-    def set(self, value=None, min_val=None, max_val=None, reverse=None, range_mode=None):
+        self._listener = []
+
+    def set(self, value=None, min_val=None,
+            max_val=None, reverse=None, range_mode=None):
         # disable DT and CLK pin interrupts
         self._hal_disable_irq()
-        
-        if value != None:
+
+        if value is not None:
             self._value = value
-        if min_val != None:
+        if min_val is not None:
             self._min_val = min_val
-        if max_val != None:
+        if max_val is not None:
             self._max_val = max_val
-        if reverse != None:
+        if reverse is not None:
             self._reverse = -1 if reverse else 1
-        if range_mode != None:
+        if range_mode is not None:
             self._range_mode = range_mode
         self._state = _R_START
 
         # enable DT and CLK pin interrupts
         self._hal_enable_irq()
-    
+
     def value(self):
         return self._value
 
     def reset(self):
         self._value = 0
-        
+
     def close(self):
         self._hal_close()
-        
+
+    def add_listener(self, l):
+        self._listener.append(l)
+
     def _process_rotary_pins(self, pin):
-        clk_dt_pins = (self._hal_get_clk_value() << 1) | self._hal_get_dt_value()
+        old_value = self._value
+        clk_dt_pins = (self._hal_get_clk_value() <<
+                       1) | self._hal_get_dt_value()
         # Determine next state
         self._state = _transition_table[self._state & _STATE_MASK][clk_dt_pins]
         direction = self._state & _DIR_MASK
-        
-        incr = 0        
+
+        incr = 0
         if direction == _DIR_CW:
             incr = 1
         elif direction == _DIR_CCW:
-            incr  = -1
-            
+            incr = -1
+
         incr *= self._reverse
-            
+
         if self._range_mode == self.RANGE_WRAP:
-            self._value = _wrap(self._value, incr, self._min_val, self._max_val)
+            self._value = _wrap(
+                self._value,
+                incr,
+                self._min_val,
+                self._max_val)
         elif self._range_mode == self.RANGE_BOUNDED:
-            self._value = _bound(self._value, incr, self._min_val, self._max_val)
+            self._value = _bound(
+                self._value,
+                incr,
+                self._min_val,
+                self._max_val)
         else:
             self._value = self._value + incr
+
+        try:
+            if old_value != self._value and len(self._listener) != 0:
+                micropython.schedule(_trigger, self)
```

Comment:
I like how you used `micropython.schedule`.  I think that will avoid any issues of allocating memory from the heap in an IRQ.